daemon/src/device.rs

```diff
@@ -74,6 +78,57 @@ struct ButtonState {
     hold_handled: bool,
 }
 
+#[derive(Debug, Clone, Default)]
+struct TapTempoState {
+    taps: Vec<Instant>,
+}
+
+impl TapTempoState {
+    fn reset_start(&mut self, now: Instant) {
+        self.taps.clear();
+        self.taps.push(now);
+    }
+
+    fn last(&self) -> Option<Instant> {
+        self.taps.last().copied()
+    }
+
+    fn register(&mut self, now: Instant, window: Duration) -> Option<Vec<Duration>> {
+        // If this is the first tap or we've exceeded the window, start fresh
+        if self.taps.is_empty() {
+            self.reset_start(now);
+            return None;
+        }
+
+        if let Some(last) = self.last() {
+            if now.duration_since(last) > window {
+                self.reset_start(now);
+                return None;
+            }
+        }
+
+        // Add the new tap
+        self.taps.push(now);
```

Comment:
This seems a little weird, would it not be better to create the taps vec with `Vec::with_capacity(4)`, then

```rust
if self.taps.length() == self.taps.capacity() {
  self.taps.remove(0);
}
taps.push(now)
```

This would allow a continuous number of taps without limit, or overflow checking, and you can work on the data available.